src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -116,6 +149,67 @@ RouterExecutorStart(QueryDesc *queryDesc, int eflags, Task *task)
 
 
 /*
+ * InitTransactionStateForTask is called during executor start with the first
+ * modifying (INSERT/UPDATE/DELETE) task during a transaction. It creates the
+ * transaction participant hash, opens connections to this task's nodes, and
+ * populates the hash with those connections after sending BEGIN commands to
+ * each. If a node fails to respond, its connection is set to NULL to prevent
+ * further interaction with it during the transaction.
+ */
+static void
+InitTransactionStateForTask(Task *task)
+{
+	MemoryContext oldContext = NULL;
+	ListCell *placementCell = NULL;
+
+	xactParticipantHash = CreateXactParticipantHash();
+
+	oldContext = MemoryContextSwitchTo(TopTransactionContext);
+
+	foreach(placementCell, task->taskPlacementList)
+	{
+		ShardPlacement *placement = (ShardPlacement *) lfirst(placementCell);
+		NodeConnectionKey participantKey;
+		NodeConnectionEntry *participantEntry = NULL;
+		bool entryFound = false;
+
+		PGconn *connection = NULL;
+
+		MemSet(&participantKey, 0, sizeof(participantKey));
+		strlcpy(participantKey.nodeName, placement->nodeName,
+				MAX_NODE_LENGTH + 1);
+		participantKey.nodePort = placement->nodePort;
+
+		participantEntry = hash_search(xactParticipantHash, &participantKey,
+									   HASH_ENTER, &entryFound);
+		Assert(!entryFound);
+
+		connection = GetOrEstablishConnection(placement->nodeName,
+											  placement->nodePort);
+		if (connection != NULL)
+		{
+			PGresult *result = PQexec(connection, "BEGIN");
+			if (PQresultStatus(result) != PGRES_COMMAND_OK)
+			{
+				WarnRemoteError(connection, result);
+				PurgeConnection(connection);
+
+				connection = NULL;
+			}
+
+			PQclear(result);
+		}
```

Comment:
> Hm. Should we error out if we reach the `connection == NULL` case here?

No. A `NULL` connection just means that node has stopped being able to participate in the transaction, for whatever reason. A single node not being able to `BEGIN` isn't a reason to completely abort. I suppose if _all_ nodes could not start the transaction, we could abort here, but it's easier to just set all their connections to `NULL` and let the `Run` function error out (it will say "unable to modify any active placements").